File: go-sdk/crypto.go
package receipts

import (
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"time"
)

// SHA256Hex returns the lowercase hex SHA-256 of bytes.
func SHA256Hex(b []byte) string {
	h := sha256.Sum256(b)
	return hex.EncodeToString(h[:])
}

// KeyPair mirrors the JSON shape used by the TypeScript SDK.
type KeyPair struct {
	Kid        string `json:"kid"`
	PublicKey  string `json:"public_key"`  // base64
	PrivateKey string `json:"private_key"` // base64 (seed, 32 bytes)
	Algorithm  string `json:"algorithm"`
	Curve      string `json:"curve"`
	CreatedAt  string `json:"created_at"`
}

// GenerateKeyPair creates a fresh Ed25519 keypair. The private_key field
// is the 32-byte SEED (not the 64-byte expanded private key) — this
// matches the wire format used by @noble/ed25519 v2 and the Python SDK.
func GenerateKeyPair() (KeyPair, error) {
	seed := make([]byte, ed25519.SeedSize)
	if _, err := rand.Read(seed); err != nil {
		return KeyPair{}, err
	}
	priv := ed25519.NewKeyFromSeed(seed)
	pub := priv.Public().(ed25519.PublicKey)
	suffix := make([]byte, 6)
	if _, err := rand.Read(suffix); err != nil {
		return KeyPair{}, err
	}
	return KeyPair{
		Kid:        "go-" + hex.EncodeToString(suffix),
		PublicKey:  base64.StdEncoding.EncodeToString(pub),
		PrivateKey: base64.StdEncoding.EncodeToString(seed),
		Algorithm:  "EdDSA",
		Curve:      "ed25519",
		CreatedAt:  time.Now().UTC().Format("2006-01-02T15:04:05.000Z"),
	}, nil
}

// Sign signs payload with the keypair's private key (seed).
func Sign(payload []byte, kp KeyPair) (string, error) {
	seed, err := base64.StdEncoding.DecodeString(kp.PrivateKey)
	if err != nil {
		return "", err
	}
	if len(seed) != ed25519.SeedSize {
		return "", errors.New("private_key must decode to 32 bytes (Ed25519 seed)")
	}
	priv := ed25519.NewKeyFromSeed(seed)
	sig := ed25519.Sign(priv, payload)
	return base64.StdEncoding.EncodeToString(sig), nil
}

// Verify returns true iff the signature is valid.
func Verify(payload []byte, signatureB64 string, publicKeyB64 string) bool {
	sig, err := base64.StdEncoding.DecodeString(signatureB64)
	if err != nil || len(sig) != ed25519.SignatureSize {
		return false
	}
	pub, err := base64.StdEncoding.DecodeString(publicKeyB64)
	if err != nil || len(pub) != ed25519.PublicKeySize {
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(pub), payload, sig)
}

File: rust-sdk/src/crypto.rs
//! Ed25519 + SHA-256 primitives.

use base64::{engine::general_purpose::STANDARD, Engine};
use chrono::Utc;
use ed25519_dalek::{Signature, Signer, SigningKey, Verifier, VerifyingKey};
use rand_core::OsRng;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

/// Compute SHA-256 of bytes, return lowercase hex.
pub fn sha256_hex(b: &[u8]) -> String {
    let h = Sha256::digest(b);
    hex::encode(h)
}

/// Wire-compatible KeyPair (matches TS/Python/Go shape).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KeyPair {
    pub kid: String,
    pub public_key: String,
    pub private_key: String,
    pub algorithm: String,
    pub curve: String,
    pub created_at: String,
}

pub fn generate_keypair() -> KeyPair {
    let mut csprng = OsRng;
    let sk = SigningKey::generate(&mut csprng);
    let vk = sk.verifying_key();

    // 6-byte kid suffix
    let mut suffix = [0u8; 6];
    use rand_core::RngCore;
    OsRng.fill_bytes(&mut suffix);
    let kid = format!("rs-{}", hex::encode(suffix));

    KeyPair {
        kid,
        public_key: STANDARD.encode(vk.as_bytes()),
        private_key: STANDARD.encode(sk.to_bytes()),
        algorithm: "EdDSA".to_string(),
        curve: "ed25519".to_string(),
        created_at: Utc::now()
            .format("%Y-%m-%dT%H:%M:%S%.3fZ")
            .to_string(),
    }
}

pub fn sign(payload: &[u8], kp: &KeyPair) -> Result<String, String> {
    let seed = STANDARD
        .decode(&kp.private_key)
        .map_err(|e| format!("private_key decode: {e}"))?;
    if seed.len() != 32 {
        return Err(format!(
            "private_key must decode to 32 bytes (got {})",
            seed.len()
        ));
    }
    let mut arr = [0u8; 32];
    arr.copy_from_slice(&seed);
    let sk = SigningKey::from_bytes(&arr);
    let sig: Signature = sk.sign(payload);
    Ok(STANDARD.encode(sig.to_bytes()))
}

pub fn verify(payload: &[u8], signature_b64: &str, public_key_b64: &str) -> bool {
    let sig = match STANDARD.decode(signature_b64) {
        Ok(v) if v.len() == 64 => {
            let mut a = [0u8; 64];
            a.copy_from_slice(&v);
            Signature::from_bytes(&a)
        }
        _ => return false,
    };
    let pk = match STANDARD.decode(public_key_b64) {
        Ok(v) if v.len() == 32 => {
            let mut a = [0u8; 32];
            a.copy_from_slice(&v);
            match VerifyingKey::from_bytes(&a) {
                Ok(v) => v,
                Err(_) => return false,
            }
        }
        _ => return false,
    };
    pk.verify(payload, &sig).is_ok()
}
